src/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs

```diff
@@ -123,13 +123,110 @@ private static Exception CreateNamedMapsNotSupportedException()
+            string mapName = "/corefx_map_" + Guid.NewGuid().ToString("N");
+
+            // Determine the flags to use when creating the shared memory object
+            Interop.Sys.OpenFlags flags = (protections & Interop.Sys.MemoryMappedProtections.PROT_WRITE) != 0 ?
+                Interop.Sys.OpenFlags.O_RDWR :
+                Interop.Sys.OpenFlags.O_RDONLY;
+            flags |= Interop.Sys.OpenFlags.O_CREAT | Interop.Sys.OpenFlags.O_EXCL; // CreateNew
+
+            // Determine the permissions with which to create the file
+            Interop.Sys.Permissions perms = default(Interop.Sys.Permissions);
+            if ((protections & Interop.Sys.MemoryMappedProtections.PROT_READ) != 0)
+                perms |= Interop.Sys.Permissions.S_IRUSR;
+            if ((protections & Interop.Sys.MemoryMappedProtections.PROT_WRITE) != 0)
+                perms |= Interop.Sys.Permissions.S_IWUSR;
+            if ((protections & Interop.Sys.MemoryMappedProtections.PROT_EXEC) != 0)
+                perms |= Interop.Sys.Permissions.S_IXUSR;
+
+            // Create the shared memory object.
+            int fd = Interop.Sys.ShmOpen(mapName, flags, (int)perms);
+            if (fd < 0)
+            {
+                // If ShmOpen fails, fall back to file backing object.
+                // Note that the System.Native shim will force failure on platforms where
+                // the result of native shm_open does not work well with our subsequent call
+                // to mmap.
+                return null; 
+            }
+
+            SafeFileHandle fileHandle = new SafeFileHandle((IntPtr)fd, ownsHandle: true);
+            try
+            {
+                // Unlink the shared memory object immediatley so that it'll go away once all handles 
+                // to it are closed (as with opened then unlinked files, it'll remain usable via
+                // the open handles even though it's unlinked and can't be opened anew via its name).
+                Interop.CheckIo(Interop.Sys.ShmUnlink(mapName));
+
+                // Give it the right capacity.  We do this directly with ftruncate rather
+                // than via FileStream.SetLength after the FileStream is created because, on some systems,
+                // lseek fails on shared memory objects, causing the FileStream to think it's unseekable,
+                // causing it to preemptively throw from SetLength.
+                Interop.CheckIo(Interop.Sys.FTruncate(fd, capacity));
+
+                // Wrap the file descriptor in a stream and return it.
+                return new FileStream(fileHandle, TranslateProtectionsToFileAccess(protections));
+            }
+            catch
+            {
+                fileHandle.Dispose();
+                throw;
+            }
+        }
+
+        private static readonly string s_tempMapsDirectory = PersistedFiles.GetTempFeatureDirectory("maps");
```

Comment:
We should probably lazily initialize this now, since it'll only be used in the fallback case.